### DISCUSSION TITLE
SQLAlchemy / Celery: Sqlalchemy error: is already attached to session '12' (this is' 5 ')

### DISCUSSION THREAD

**Discussion by thisiseddy-ab** [sqlalchemy/sqlalchemy#8144] [Usage Questions] [answered]
I'm trying to aggregate News RSS Feeds with celery but im getting this error from SQLAlchemy.

Sqlalchemy error: is already attached to session '12' (this is' 5 ')

Im runing 3 celery workers like thi:

celery -A main.celery worker --loglevel=INFO -P gevent --concurrency=50 -n celeryworker1@%%h

From reading documention this how i think u need to setup SQLAlchemy for mutithreading with Scoped Session.

Databse Settings:

```py
        engine = create_engine(SQLALCHEMY_DATABASE_URL,pool_size=200, max_overflow=250)
 
        SessionLocal = sessionmaker(bind=engine,autocommit=False, autoflush=False)

        Session_t = scoped_session(SessionLocal)
```

Session Settings:
```py
  @contextmanager
  def session_scope():
      session = Session_t()
      try:
          yield session
      finally:
          Session_t.remove()
```
The Problem is when i create News_Article_Tag object, the problem is so news articles have same tages and are created at the same time beacause of the multithreading and concurrency.

```py
      for tag in article["tags"]:
      
          tag_query = db.query(models.News_Article_Tag).filter_by(country = newspaper.country,language = newspaper.language, tag = tag).first()
          
          if not tag_query:
              new_tag = models.News_Article_Tag(
                  country = newspaper.country,
                  language = newspaper.language,
                  tag = tag,
                  popularity = 1
                  )
      
              new_article_tags.append(new_tag)
              article_tags.append(new_tag)
          else:
              tag_query.popularity = tag_query.popularity + 1
              article_tags.append(tag_query)
      
      new_article.tags.extend(article_tags)
      db.add(new_article)
      db.add_all(new_article_tags)
      db.commit()
```
I try it to solved by checking if object is in session like this.

```py
            for tag in article["tags"]:
            
                tag_query = db.query(models.News_Article_Tag).filter_by(country = newspaper.country,language = newspaper.language, tag = tag).first()
                
                if not tag_query:
                    new_tag = models.News_Article_Tag(
                        country = newspaper.country,
                        language = newspaper.language,
                        tag = tag,
                        popularity = 1
                        )
                    article_tags.append(new_tag)
            
                    if new_tag not in session:
                        session.add(new_tag)
                        session.commit()
                    
                else:
                    tag_query.popularity = tag_query.popularity + 1
                    article_tags.append(tag_query)
            
            new_article.tags.extend(article_tags)
            session.add(new_article)
            session.commit()
```
But Still does not work, i think it checks just in current session and not sessions in other threads.

Does someone has an idea how to solve it ?

Is there a way to get all sessions to check if they have this object ?

Or Any Solution is Welcomed

I don't know is this what im looking for (ScopedRegistry) ?

https://docs.sqlalchemy.org/en/14/orm/contextual.html#sqlalchemy.util.ScopedRegistry

**Comment by CaselIT** [comment]
Hi,

could you reformat the code properly? Otherwise it's very hard to read it. https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

**Reply by thisiseddy-ab** [reply]
im trying i put the code in blocks it just show it right, im new at github

here is the stackoverflow question, there is properly formated
https://stackoverflow.com/questions/72646524/sqlalchemy-celery-sqlalchemy-error-is-already-attached-to-session-12-this

**Comment by CaselIT** [comment, marked as answer]
Hi,

I've formatted the text above.

Please note that geven are not threads.
If you use the `def session_scope():` you showed, I think you can avoid using `scoped_session` that by default will not work with gevents.

You have to remember to create a new session with your `session_scope` for each task invocation. Somethig like this
```py
engine = create_engine(SQLALCHEMY_DATABASE_URL,pool_size=200, max_overflow=250)
SessionLocal = sessionmaker(bind=engine,autocommit=False, autoflush=False)

@contextmanager
def session_scope():
    session = SessionLocal()
    try:
        yield session
    finally:
      session.close()

@app.task
def some_task(seme_input):
  with session_scope() as session:
    ... # use the session
```

**Reply by thisiseddy-ab** [reply]
i try it first  with out scoped session i was getting alot errors, i will try it again i will see how it goes

**Reply by thisiseddy-ab** [reply]
it try it now, like you said and it works fine, one question i have does this method creates a postgress connection per worker, or many depending from concurency ?

If it is one connection per worker is great, i can scale it up more workers

**Reply by CaselIT** [reply]
I think it will depend on the concurrency. If a worker that uses gthread has 10 task running at the same time it will use 10 connections (if each task creates a session that is, if you have tasks that don't use the db they don't count in the total above).

**Reply by thisiseddy-ab** [reply]
yes your right i tested it, every concurrent task creates a connection.Of Course if i use db

**Reply by CaselIT** [reply]
glad that worked for you!
